tokio-timer/src/debounce.rs

```diff
@@ -0,0 +1,408 @@
+//! Debounce streams on the leading or trailing edge or both edges for a certain
+//! amount of time.
+//!
+//! See [`Debounce`] for more details.
+//!
+//! [`Debounce`]: struct.Debounce.html
+
+use {clock, Delay, Error};
+
+use futures::{
+    future::Either,
+    prelude::*,
+};
+use std::{
+    cmp,
+    error::Error as StdError,
+    fmt::{Display, Formatter, Result as FmtResult},
+    time::{Duration, Instant},
+};
+
+/// Debounce streams on the leading or trailing edge or both.
+///
+/// Useful for slowing processing of e. g. user input or network events
+/// to a bearable rate.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct Debounce<T: Stream> {
+    delay: Option<Delay>,
+    duration: Duration,
+    edge: Edge,
+    last_item: Option<T::Item>,
+    max_wait: Option<Duration>,
+    max_wait_to: Option<Instant>,
+    stream: T,
+}
+
+/// Builds a debouncing stream.
+#[derive(Debug)]
+pub struct DebounceBuilder<T> {
```

Comment:
This should be named `Builder` and be referred to `debounce::Builder`. In general, idiomatic Rust tries to avoid stuttering (debounce::DebounceBuilder).